apps/bible/models.py

```diff
@@ -238,11 +238,19 @@ def from_reference(cls, reference, persist=False):
             verse_end_separator = match.group('verse_end_separator')
             verse_end = match.group('verse_end')
             verse_end = int(verse_end) if verse_end else None
+            has_ff = bool(match.group('ff'))
```

Comment:
How about we use a recursive approach?

```suggestion
            has_ff = bool(match.group('ff'))
            
            if has_ff:
                # resolve the passage recursively
                non_abbreviated_passage = cls.from_reference(reference.removesuffix('ff'), persist=False)
                verse_end = non_abbreviated_passage.chapter_end.bibleverses.last().verse
            else:
                # NOOP: reference does not contain the "following" (`ff`) abbreviation
                pass
```

And remove all the other code changes in this function?

Reply:
Thank you. I only updated the logic for the case when `chapter_end` is `None`, such as John 3:16ff.